docs/using.md

```diff
@@ -171,41 +489,294 @@ is passed in `infrastructure_provisioning: bar` and the third `arbitrary_key: ba
+-   For a list of values, $eq evaluates to true if it is equal to at least one.
+-   For a list of values, $neq evaluates to true if it is equal to none of the values.
+-   ThenRun blocks are optional.
+    -   ****Most usecases do not need to use ThenRun****
+    -   If you do use ThenRun, please be judicious. If you have a task that is scheduled when
+        mongodb<sub>setup</sub> == replica, it would be confusing if mongodb<sub>setup</sub> was overwritten to standalone.
+        But it would be ok to overwrite mongodb<sub>setup</sub> to replica-delay-mixed, as is done in the
+        [ParallelWorkload][pi] workload.
+-   Each item in the ThenRun list can only support one {bootstrap<sub>key</sub>: bootstrap<sub>value</sub>} pair.
+-   If using ThenRun but you would also like to schedule a task without any bootstrap overrides,
+    Add an extra pair to ThenRun with the original key/value, like done on line 189 [here][pi].
+-   If using ThenRun, the new task name becomes <taskname>\_<bootstrap-value>. In the ParallelWorkload example,
+    the task name becomes parallel<sub>insert</sub><sub>replica</sub><sub>delay</sub><sub>mixed</sub> (name is automatically converted to snake<sub>case</sub>).
+    The bootstrap-key is not included in the name for the purpose of not changing existing names and
+    thus deleting history. This may change after PM-2310.
+
+
+<a id="orgd89f221"></a>
+
+## Generators
+
+It is often necessary to use Genny to operate with large amounts of data which would be impractical to hardcode. Genny uses generators for this. A generator is a piece of code that generates pseudorandom values every time it is invoked, and which can be configured from the workload yaml. Notably, generators use a hardcoded seed, which is always automatically reused by default, so repeated Genny executions should be deterministic with respect to generated values. A user wanting to vary the generated docs can vary the seed. See [./src/workloads/docs/GeneratorsSeeded.yml](../src/workloads/docs/GeneratorsSeeded.yml) for an example.
+
+Generators are not a builtin feature of Genny, but must be integrated by each actor for the configuration values that accept them. For examples of using generators, see [./src/workloads/docs/Generators.yml](../src/workloads/docs/Generators.yml). To integrate generators into an actor, use the [DocumentGenerator](../src/value_generators/include/value_generators/DocumentGenerator.hpp) with the yaml node you intend to generate documents from. (And see [here](#org7e6c6bd) for more details on creating an actor in the first place.)
+
+
+<a id="org2078b23"></a>
+
+## Preprocessor
+
+For convenience when developing workloads, Genny offers a preprocessing syntax that can be used for configuration reuse and parameterization. Remember: when developing a workload, you can always check results of preprocessing:
+
+    ./run-genny evaluate src/workloads/<workload_dir/workload_name.yml>
+	
+This command helps find yaml-based mistakes.
```

Comment:
Somewhere add a caveat that we don't do schema checking of the yaml - if you fat-finger a dictionary key and an Actor or something never uses it and instead uses a default there's no protection against that. Perhaps a troubleshooting or word-to-the-wise or something and perhaps link to the ticket to add a ["strict mode"](https://jira.mongodb.org/browse/TIG-2766).